### PR TITLE
Add XML docs to some Linq iterators

### PR DESCRIPTION
**Pull request by jamesqo**
This PR adds XML comment documentation to many of the manual (class-based) iterators we have in Linq. I have not touched the `IGrouping` / `IOrderedEnumerable` related stuff because I haven't used that API too much yet.

@JonHanna @stephentoub PTAL

### CODE REVIEW

**Comment by JonHanna** [comment]
Looks good at a quick read. I'll take a more thorough look this afternoon Irish time.

File: src/System.Linq/src/System/Linq/Union.cs [hunk 1]
```diff
@@ -31,6 +31,10 @@ public static IEnumerable<TSource> Union<TSource>(this IEnumerable<TSource> firs
             return union != null && AreEqualityComparersEqual(comparer, union._comparer) ? union.Union(second) : new UnionIterator2<TSource>(first, second, comparer);
         }
 
+        /// <summary>
+        /// An iterator that shares code among iterators returned by <see cref="Enumerable.Union"/>.
```
Comment (JonHanna) [inline src/System.Linq/src/System/Linq/Union.cs]: As with AppendPrepend on description of abstract classes.

**Comment by karelz** [comment]
Just a warning that as of today, it is unclear if/how XML docs in CoreFX repo will make it into official documentation. We had a chat with docs team few weeks back. They are working on a plan and tooling. We will have to work closely with them to ensure there is reasonable flow between source code and docs repo.

**Comment by JonHanna** [comment]
These aren't exposed members, so they shouldn't be in official docs anyway.

File: src/System.Linq/src/System/Linq/Partition.cs [hunk 2]
```diff
@@ -77,8 +77,15 @@ internal interface IPartition<TElement> : IIListProvider<TElement>
         TElement TryGetLast(out bool found);
     }
 
+    /// <summary>
+    /// Represents an enumerable with zero elements.
```
Comment (jamesqo) [inline src/System.Linq/src/System/Linq/Partition.cs]: Both done 👍

**Comment by JonHanna** [comment]
It could be opined that we therefore don't need these at all, but I think this project gets enough attention that they'll be read.

File: src/System.Linq/src/System/Linq/Iterator.cs [hunk 3]
```diff
@@ -9,58 +9,93 @@ namespace System.Linq
 {
     public static partial class Enumerable
     {
+        /// <summary>
+        /// A base class for enumerables that are loaded on-demand.
+        /// </summary>
+        /// <typeparam name="TSource">The type of each item to yield.</typeparam>
+        /// <remarks>
+        /// - The value of an iterator is immutable; when re-enumerated, it should always yield
```
Comment (JonHanna) [inline src/System.Linq/src/System/Linq/Iterator.cs]: The really nice one that isn't on that page is `<inheritdoc/>`. Unfortunately the reason it isn't is that most of the IDEs (including VisualStudio) don't pay attention to it for intellisense, but SandCastle does so it's handy as far as generated documentation goes.

**Comment by jamesqo** [comment]
@karelz 

> Just a warning that as of today, it is unclear if/how/when XML docs in CoreFX repo will make it into official documentation. We had a chat with docs team few weeks back. They are working on a plan and tooling. We will have to work closely with them to ensure there is reasonable flow between source code and docs repo.

Unrelated to this PR, but is there such a plan for CoreCLR as well? Is the docs team planning to insert XML documentation in S.P.CoreLib?

**Comment by karelz** [comment]
I expect the plan will apply to both CoreFX and CoreCLR repos. At least I don't see any reason why not at this moment. I will make sure we have a good story for contributors :)

**Comment by jamesqo** [comment]
Sounds great! :smile: Feel free to update this thread (or ping me in some related issue) when that happens.

**Comment by karelz** [comment]
Sure, it is still month(s) out.

**Comment by danmosemsft** [comment]
@VSadov would you like to hit merge?

**Pull request merged**